language/documentation/hackathon-tutorial/step_7/BasicCoin/sources/BasicCoin.move

```diff
@@ -0,0 +1,94 @@
+/// This module defines a minimal and generic Coin and Balance.
+module NamedAddr::BasicCoin {
+    use Std::Errors;
+    use Std::Signer;
+
+    /// Error codes
+    const ENOT_MODULE_OWNER: u64 = 0;
+    const EINSUFFICIENT_BALANCE: u64 = 1;
+    const EALREADY_HAS_BALANCE: u64 = 2;
+
+    struct Coin<phantom CoinType> has store {
+        value: u64
+    }
+
+    struct Balance<phantom CoinType> has key {
+        coin: Coin<CoinType>
+    }
+
+    /// Publish an empty balance resource under `account`'s address. This function must be called before
+    /// minting or transferring to the account.
+    public fun publish_balance<CoinType>(account: &signer) {
+        let empty_coin = Coin<CoinType> { value: 0 };
+        assert!(!exists<Balance<CoinType>>(Signer::address_of(account)), Errors::already_published(EALREADY_HAS_BALANCE));
+        move_to(account, Balance<CoinType> { coin:  empty_coin });
+    }
+
+    /// Mint `amount` tokens to `mint_addr`. This method requires a witness with `CoinType` so that the
+    /// module that owns `CoinType` can decide the minting policy.
+    public fun mint<CoinType: drop>(mint_addr: address, amount: u64, _witness: CoinType) acquires Balance {
+        // Deposit `total_value` amount of tokens to mint_addr's balance
+        deposit(mint_addr, Coin<CoinType> { value: amount });
+    }
+
+    public fun balance_of<CoinType>(owner: address): u64 acquires Balance {
+        borrow_global<Balance<CoinType>>(owner).coin.value
+    }
+
+    /// Transfers `amount` of tokens from `from` to `to`. This method requires a witness with `CoinType` so that the
+    /// module that owns `CoinType` can decide the transferring policy.
+    public fun transfer<CoinType: drop>(from: &signer, to: address, amount: u64, _witness: CoinType) acquires Balance {
+        let check = withdraw<CoinType>(Signer::address_of(from), amount);
+        deposit<CoinType>(to, check);
+    }
+
+    spec transfer {
```

Comment:
Should discuss whether these specs are too complicated. You have little chance to discuss all this in like 10 minutes.

Alternative: use one of the previous examples with missing asserts, and show first verification without any specs. The prover should show all code positions where an abort with EXECUTION_FAILURE happens. Then perhaps add asserts and aborts_if. No ensures or perhaps only one as an example.

Also focus on perhaps one function (transfer).

Also specs can be significantly simplifier if addr_from != to is a requirement for transfer (assert). You can even make a point of first adding the trivial ensures, show they fail, and then add the assert/aborts_if.